### ISSUE TITLE
SC2034: Allow _xyz to be considered throwaway variable

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#1498]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
- My shellcheck version (`shellcheck --version` or "online"): N/A
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

_More of a discussion ticket rather than an explicit feature request._

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

_comp1() {
	local cur _prev opts
	COMPREPLY=()
	cur="${COMP_WORDS[COMP_CWORD]}"
	_prev="${COMP_WORDS[COMP_CWORD-1]}"
	opts="a b c"

	mapfile -t COMPREPLY <<< "$(compgen -W "${opts}" -- "${cur}")"
	return 0
}
complete -F _comp1 comp1
```

#### Here's what shellcheck currently says:

```bash
<untitled 1744>:
  21:2   warning  shellcheck :warning  _prev appears unused. Verify use (or export if used externally). [SC2034]
```

#### Here's what I wanted or expected to see:

Nothing. A lot of languages / IDEs consider that status quo.
Would it make sense to apply the same logic here too?

I understand that e.g. `_comp1` is one style of specifying `COMPREPLY` functions, so, I am not sure how viable would that change be

**Comment by koalaman** [comment]
This came up once before in #1193. Which languages and IDEs use this convention?

**Comment by stdedos** [comment]
JetBrain's IDEs use this on their linter (where applicable)

Ruby:
> Update: In Ruby 2.0, you can use any variable starting with _ to signify it is unused. This way the variable name can be more explicit about what is being ignored
> https://stackoverflow.com/a/8691749/2309247

Python:
> In Python _ is definitively acceptable. It can however conflict with gettext alias  _().
> https://docs.python.org/3/library/gettext.html

Perl:
> This would be bad in perl, which uses $_ (and sometimes _) as a special variable.
> https://softwareengineering.stackexchange.com/a/147326/329481

-------

I cannot speak "official" things, but here is some "internet wisdom"

**Comment by benesch** [comment]
FWIW, the leading-underscore convention is also used by the Rust compiler to determine whether a variable is intentionally unused: https://stackoverflow.com/questions/48361537/why-do-underscore-prefixed-variables-exist.

**Comment by Nightfirecat** [comment]
I'm not sure what Rust has to do with this pattern specifically, many languages encourage this type of naming for unused variables.

**Comment by benesch** [comment]
Yes. @koalaman was asking for specific examples of languages that support this pattern earlier in this thread:

> This came up once before in #1193. Which languages and IDEs use this convention?

**Comment by Nightfirecat** [comment]
Ah my bad, I missed that comment. 👍

**Comment by alkisg** [comment]
Hi, I came here to file the exact same suggestion. So just +1, to increase its popularity. :)

[The docs](https://github.com/koalaman/shellcheck/wiki/SC2034) suggest using _, but this is used very frequently as a gettext alias, so I wouldn't like to use it for throwaway variables; and additionally, `read -r _name _surname age` is more descriptive than `read -r _ _ age`.

Btw thank you for shellcheck, it's an awesome tool!

**Comment by stdedos** [comment]
@koalaman If I poke gently, will you re-consider implementing this?

**Comment by vatosarmat** [comment]
One of the most popular languages nowadays - Typescript have [that feature](https://www.typescriptlang.org/docs/handbook/release-notes/overview.html#destructured-variables-can-be-explicitly-marked-as-unused)

**Comment by RomanValov** [comment]
Another use case for this is kind of library script file to be sourced by several other scripts.
I would like to put some shared variables in my library script. But not all other scripts are using variables:

```
# lib.sh

_USE_FEATURE_1=true
_USE_FEATURE_2=false
```

```
# script1.sh

source lib.sh

if "$_USE_FEATURE_1"; then
  ...
fi
```

```
# script2.sh

source lib.sh

if "$_USE_FEATURE_2"; then
  ...
fi
```

So it would be nice to have some kind of convention to define those optionally used variables without special shellcheck directives in the code.

**Comment by FrankHB** [comment]
As a note, I'd add a bash-specific use case here.

Consider `: {var:=$(command)}` to initialize `var` if not set or empty, since `:` always has a zero exit status, `set -e` will not cause the shell to exit on failure of `command`. This may be quite unexpected. A workaround is `var={var:-$(command)}`, but this is verbose when `var` is a long name. Then there's the workaround: `_={var:=$(command)}`. The special variable `_` in bash is the last argument of the last command, so if the command succeeds, the meaning is perfectly compatible (to one-argument cases of `:` above).

I find it by myself and don't know whether this is an established idiom, but it seems fit to the wiki page to mention. However, there is another issue: assignment to `_` may lead to shell-specific behavior. @koalaman, is this also worth mentioning in the wiki page, or a standalone warning?
